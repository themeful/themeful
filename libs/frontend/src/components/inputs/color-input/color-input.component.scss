:host {
  display: block;
}

.color-input {
  display: block;
  color: var(--tf-font-color);

  &__label {
    font: var(--tf-small-font);
    width: 100%;
  }

  &__hint {
    font: var(--tf-small-font);
    display: block;
    min-height: 0.875rem;
    width: 100%;
    color: #ff6464;
    margin: 1px 0 4px;
  }

  &__input {
    font: var(--tf-default-font);
    width: 100%;
    color: var(--tf-font-color);
    border: 0;
    background-color: rgba(128, 128, 128, 0.1);
    border-bottom: 1px solid currentColor;
    box-sizing: border-box;

    &:focus {
      outline: 0 !important;
      border-color: #325af0;
    }
  }

  &--error &__input {
    border-color: #ff6464;
  }

  &--valid &__input {
    border-color: #23c223;
  }

  &__control-wrapper {
    margin-bottom: 18px;
  }

  &__control-wrapper,
  &__controls {
    user-select: none;
    display: flex;
    flex-direction: column;
    gap: 8px;
  }

  &__controls {
    flex-grow: 4;
  }

  &__control-row {
    display: flex;
    flex-direction: row;
    gap: 8px;
  }

  &__selected-color {
    position: relative;
    height: 50px;
    width: 50px;
    border: 1px solid #5555;
    border-radius: 2px;
    overflow: hidden;
    box-sizing: border-box;
    z-index: 1;

    &:before,
    &:after {
      content: '';
      position: absolute;
      width: 100%;
      height: 100%;
      z-index: -1;
    }

    &:after {
      background: var(--tf-result-color);
    }

    &:before {
      background: repeating-conic-gradient(#aaa 0% 25%, #fff 0% 50%) 50% / 10px 10px;
    }

    .color-input__crossed-out {
      background-color: red;
      width: 150%;
      height: 4px;
      position: absolute;
      display: none;
      transform: translate(50%, -50%) rotate(-45deg);
      top: 50%;
      right: 50%;
    }

    &.no-color {
      .color-input__crossed-out {
        display: block;
      }

      &:before,
      &:after {
        background: none;
      }
    }
  }

  &__shade {
    position: relative;
    height: 150px;
    width: 100%;
    cursor: crosshair;
    background-color: var(--tf-hue-color);
    background-image: linear-gradient(0, #000000 0, transparent 100%),
      linear-gradient(90deg, #ffffff 0, transparent 100%);
    overflow: hidden;
  }

  &__pointer {
    pointer-events: none;
    position: absolute;
    width: 14px;
    height: 14px;
    border-radius: 50%;
    border: 2px solid #fff;
    box-shadow: 1px 1px 1px 0 rgba(0, 0, 0, 0.3);
    margin: -9px;
  }

  &__hue {
    -webkit-appearance: none;
    appearance: none;
    margin: 0;
    height: 20px;
    width: 100%;
    background: linear-gradient(
      to right,
      #ff0000 0%,
      #ffff00 17%,
      #00ff00 33%,
      #00ffff 50%,
      #0000ff 67%,
      #ff00ff 83%,
      #ff0000 100%
    );
    --thumb-color: var(--tf-hue-color);
  }

  &__alpha {
    -webkit-appearance: none;
    appearance: none;
    margin: 0;
    height: 20px;
    width: 100%;
    position: relative;
    background-color: transparent;
    z-index: 1;
    --thumb-color: var(--tf-result-color);

    &:before,
    &:after {
      content: '';
      position: absolute;
      width: 100%;
      height: 100%;
      z-index: -1;
    }

    &:after {
      background-image: linear-gradient(90deg, var(--tf-opaque-color) 0, transparent 100%);
    }

    &:before {
      background: repeating-conic-gradient(#aaa 0% 25%, #fff 0% 50%) 50% / 10px 10px;
    }
  }

  &__hue,
  &__alpha {
    &::-webkit-slider-thumb {
      -webkit-appearance: none;
      appearance: none;
      border: 1px solid #fff;
      box-shadow: 1px 1px 1px 0 #000000;
      width: 5px;
      height: 25px;
      background-color: var(--thumb-color);
      cursor: pointer;
    }

    &::-moz-range-thumb {
      border: 1px solid #fff;
      width: 5px;
      height: 25px;
      background-color: var(--thumb-color);
      cursor: pointer;
    }
  }
}
