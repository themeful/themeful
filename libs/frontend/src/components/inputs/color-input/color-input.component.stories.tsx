import { h } from '@stencil/core'
import { action } from '@storybook/addon-actions'
import Fragment from 'stencil-fragment'

export default {
  title: 'Components/Inputs/Color Input',
  args: {
    samples: [
      {
        label: 'Empty',
        required: true,
        format: 'HSL',
      },
      {
        label: 'HEX',
        value: '#d920c1',
        required: true,
        format: 'HEX',
      },
      {
        label: 'RGBA',
        value: 'rgba(45,79,39,0.76)',
        format: 'RGB',
      },
      {
        label: 'Invalid',
        value: 'D920C',
      },
    ],
  },
}

export const colorInput = ({ samples }) => {
  return (
    <div>
      <div class="header">
        <h1>Color Input</h1>
      </div>
      <div
        class="content"
        style={{
          display: 'grid',
          gridTemplateColumns: '300px 300px',
          gridAutoRows: '300px',
        }}
      >
        {samples.map((args) => (
          <Fragment>
            <div class="tf-light" style={{ background: '#fff', padding: '1rem' }}>
              <tf-color-input {...args} onInputChange={action('change')} />
            </div>
            <div style={{ padding: '1rem' }}>
              <tf-color-input {...args} onInputChange={action('change')} />
            </div>
          </Fragment>
        ))}
      </div>
    </div>
  )
}
