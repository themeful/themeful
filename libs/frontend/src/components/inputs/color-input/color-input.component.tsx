import { Component, Element, Event, EventEmitter, h, Host, Method, Prop } from '@stencil/core'
import { ColorTranslator } from 'colortranslator'
import { ColorInput } from 'colortranslator/dist/@types'
import {
  combineLatest,
  debounceTime,
  distinctUntilChanged,
  map,
  sampleTime,
  Subject,
  Subscription,
  tap,
} from 'rxjs'

type ColorFormat = 'HEX' | 'HSL' | 'RGB'

@Component({
  tag: 'tf-color-input',
  styleUrl: 'color-input.component.scss',
  shadow: true,
})
export class ColorInputComponent {
  private input: HTMLInputElement
  private alpha: HTMLInputElement
  private hue: HTMLInputElement
  private pointer: HTMLElement
  private trackMouseMove = false
  private tileOffset = { x: 0, y: 0 }
  private changeSource: 'input' | 'controls' = 'input'

  /** Element */
  @Element() el: HTMLTfColorInputElement

  /** Input label */
  @Prop() label: string

  /** Required input */
  @Prop() required = false

  /** Input value */
  @Prop({ mutable: true }) value: string

  /** Input Event */
  @Event({ composed: false }) inputChange: EventEmitter

  private touched = false
  private changed = false
  private valid: boolean
  private error = ''

  /** Validate value */
  @Method()
  public validate(): Promise<boolean> {
    this.touched = true
    this.value = this.value || this.input.value
    return Promise.resolve(this.internalValidation())
  }

  /** Changed value */
  @Method()
  public dirty(): Promise<boolean> {
    return Promise.resolve(this.changed)
  }

  private format: ColorFormat = 'HEX'
  private hue$ = new Subject()
  private alpha$ = new Subject()
  private shade$ = new Subject()
  private input$ = new Subject()
  private sub = new Subscription()

  private validateColor(color: string): boolean {
    try {
      new ColorTranslator(color)
      return true
    } catch {
      return false
    }
  }

  private setControls(newColor): void {
    const color = new ColorTranslator(this.valid ? newColor : 'hsl(180,50%,40%)')
    this.changeSource = 'input'
    this.input$.next(this.input.value)
    this.hue.value = `${color.H}`
    this.alpha.value = `${(1 - color.A) * 100}`
    this.alpha$.next(this.alpha.value)
    this.hue$.next(this.hue.value)
    this.shade$.next({
      x: Math.round((color.S / 100) * this.el.offsetWidth),
      y: Math.round((1 - color.L / (100 - color.S / 2)) * 150),
    })
  }

  private inputChanged = () => {
    this.changed = true
    if ((!this.valid && this.touched) || this.value !== this.input.value) {
      this.value = this.input.value
      this.internalValidation()
    }
    this.setControls(this.value)
  }

  private blur = (): void => {
    this.internalValidation()
    this.touched = true
  }

  public componentWillLoad(): void {
    this.format = this.getFormat(this.value)
    this.valid = this.validateColor(this.value)

    this.sub.add(
      this.input$.pipe(sampleTime(50), distinctUntilChanged()).subscribe((value: string) => {
        this.changed = true
        this.value = value
        this.input.value = value
      })
    )
    this.sub.add(
      this.input$.pipe(debounceTime(200), distinctUntilChanged()).subscribe((value: string) => {
        this.inputChange.emit(value)
      })
    )

    this.sub.add(
      combineLatest([
        this.hue$.pipe(
          sampleTime(25),
          distinctUntilChanged(),
          map((h: string) => parseInt(h))
        ),
        this.shade$.pipe(
          sampleTime(25),
          distinctUntilChanged(
            (prev: { x: number; y: number }, curr: { x: number; y: number }) =>
              prev.x === curr.x && prev.y === curr.y
          ),
          map(({ x, y }) => ({
            left: Math.max(0, Math.min(x, this.el.offsetWidth)),
            top: Math.max(0, Math.min(y, 150)),
          })),
          tap(({ left, top }) => {
            this.pointer.setAttribute('style', `left: ${left}px; top: ${top}px;`)
          }),
          map(({ left, top }) => {
            const s = Math.round((left / this.el.offsetWidth) * 100)
            return {
              s,
              l: Math.round((100 - s / 2) * (1 - top / 150)),
            }
          })
        ),
        this.alpha$.pipe(
          sampleTime(25),
          distinctUntilChanged(),
          map((alpha: string) => Math.round(100 - parseInt(alpha)) / 100)
        ),
      ])
        .pipe(
          map(([h, { s, l }, a]) => ({ h, s, l, a })),
          tap((color) => {
            this.el.setAttribute(
              'style',
              `--tf-hue-color: hsl(${
                color.h
              }, 100%, 50%); --tf-opaque-color: ${ColorTranslator.toHEX(
                color
              )}; --tf-result-color: ${ColorTranslator.toRGBA(color)};`
            )
          }),
          tap((values) => {
            if (this.changeSource === 'controls') {
              this.valid = true
              this.error = ''
              this.setInput(values)
            }
          })
        )
        .subscribe()
    )
  }

  public componentDidLoad(): void {
    this.setControls(this.value)
  }

  public disconnectedCallback(): void {
    this.sub.unsubscribe()
  }

  private setInput = (color: ColorInput): void => {
    const colorObj = new ColorTranslator(color)
    this.input$.next(colorObj[`${this.format}${colorObj.A < 1 ? 'A' : ''}`])
  }

  private toggleFormat = () => {
    this.format = {
      HSL: 'HEX',
      HEX: 'RGB',
      RGB: 'HSL',
    }[this.format] as ColorFormat
    if (this.valid) {
      this.setInput(this.value)
    }
  }

  private getFormat = (color = ''): ColorFormat => {
    const colorStr = color.toLowerCase()

    if (colorStr.includes('hsl')) {
      return 'HSL'
    }
    if (colorStr.includes('rgb')) {
      return 'RGB'
    }
    return 'HEX'
  }

  private trackPointer = (event): void => {
    if (this.trackMouseMove) {
      if (event.type === 'mouseup') {
        this.trackMouseMove = false
        document.removeEventListener('mousemove', this.trackPointer)
        document.addEventListener('mouseup', this.trackPointer)
      }
      this.shade$.next({
        x: event.clientX + this.tileOffset.x,
        y: event.clientY + this.tileOffset.y,
      })
    } else if (event.type === 'mousedown') {
      this.changeSource = 'controls'
      this.trackMouseMove = true
      this.tileOffset.x = event.offsetX - event.clientX
      this.tileOffset.y = event.offsetY - event.clientY
      document.addEventListener('mousemove', this.trackPointer)
      document.addEventListener('mouseup', this.trackPointer)
    }
  }

  private internalValidation = (): boolean => {
    this.error = ''
    if (this.required && (this.value === '' || this.value === undefined)) {
      this.error = `This value is required`
    } else {
      this.error = this.validateColor(this.value) ? '' : `Please enter a valid color`
    }
    this.valid = this.error === ''
    return this.valid
  }

  // =============== RENDER ===============
  private renderControls(): HTMLElement {
    return (
      <div class="color-input__control-wrapper">
        <div class="color-input__control-row">
          <div class="color-input__controls">
            <input
              class="color-input__hue"
              type="range"
              min="0"
              max="360"
              {...{
                ref: (hue: HTMLInputElement) => (this.hue = hue),
                onInput: (event) => {
                  this.changeSource = 'controls'
                  this.hue$.next((event.target as HTMLInputElement).value)
                },
              }}
            />
            <input
              class="color-input__alpha"
              type="range"
              min="0"
              max="100"
              {...{
                ref: (alpha: HTMLInputElement) => (this.alpha = alpha),
                onInput: (event) => {
                  this.changeSource = 'controls'
                  this.alpha$.next((event.target as HTMLInputElement).value)
                },
              }}
            />
          </div>
          <div
            class={`color-input__selected-color${this.valid ? '' : ' no-color'}`}
            onClick={this.toggleFormat}
            title="change format"
          >
            <div class="color-input__crossed-out"></div>
          </div>
        </div>
        <div
          {...{
            class: 'color-input__shade',
            onMouseDown: this.trackPointer,
          }}
        >
          <div
            ref={(el: HTMLInputElement) => (this.pointer = el)}
            class="color-input__pointer"
          ></div>
        </div>
      </div>
    )
  }

  public render(): HTMLTfColorInputElement {
    return (
      <Host class="color-input">
        <label
          class={`color-input${this.valid && this.touched ? ' color-input--valid' : ''}${
            !this.valid && this.touched ? ' color-input--error' : ''
          }`}
        >
          <span class="color-input__label">{this.label}</span>
          <input
            value={this.value}
            ref={(input: HTMLInputElement) => (this.input = input)}
            class="color-input__input"
            type="text"
            onInput={this.inputChanged}
            onBlur={this.blur}
          />
          <p class="color-input__hint">{this.error}</p>
        </label>
        {this.renderControls()}
      </Host>
    )
  }
}
