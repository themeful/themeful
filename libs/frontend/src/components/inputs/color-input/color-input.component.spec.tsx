import { newSpecPage } from '@stencil/core/testing'
import { ColorInputComponent } from './color-input.component'

describe('tf-color-input', () => {
  it('renders', async () => {
    const { root } = await newSpecPage({
      components: [ColorInputComponent],
      html: '<tf-color-input />',
    })
    expect(root).toEqualHtml(`
      <tf-color-input>
        <mock:shadow-root>
          <div>
            Hello, World! I'm
          </div>
        </mock:shadow-root>
      </tf-color-input>
    `)
  })

  it('renders with values', async () => {
    const { root } = await newSpecPage({
      components: [ColorInputComponent],
      html: `<tf-color-input first="Stencil" last="'Don't call me a framework' JS" />`,
    })
    expect(root).toEqualHtml(`
      <tf-color-input first="Stencil" last="'Don't call me a framework' JS">
        <mock:shadow-root>
          <div>
            Hello, World! I'm Stencil 'Don't call me a framework' JS
          </div>
        </mock:shadow-root>
      </tf-color-input>
    `)
  })
})
