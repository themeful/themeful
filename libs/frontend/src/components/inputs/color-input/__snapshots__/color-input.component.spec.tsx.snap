// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorInputComponent renders 1`] = `
<tf-color-input class="color-input">
  <mock:shadow-root>
    <label class="color-input">
      <span class="color-input__label"></span>
      <input class="color-input__input" type="text">
      <p class="color-input__hint"></p>
    </label>
    <div class="color-input__control-wrapper">
      <div class="color-input__control-row">
        <div class="color-input__controls">
          <input class="color-input__hue" max="360" min="0" type="range" value="180">
          <input class="color-input__alpha" max="100" min="0" type="range" value="0">
        </div>
        <div class="color-input__selected-color no-color" title="change format">
          <div class="color-input__crossed-out"></div>
        </div>
      </div>
      <div class="color-input__shade">
        <div class="color-input__pointer"></div>
      </div>
    </div>
  </mock:shadow-root>
</tf-color-input>
`;

exports[`ColorInputComponent renders with values 1`] = `
<tf-color-input class="color-input">
  <mock:shadow-root>
    <label class="color-input">
      <span class="color-input__label">
        Some color
      </span>
      <input class="color-input__input" type="text" value="#d920c1">
      <p class="color-input__hint"></p>
    </label>
    <div class="color-input__control-wrapper">
      <div class="color-input__control-row">
        <div class="color-input__controls">
          <input class="color-input__hue" max="360" min="0" type="range" value="308">
          <input class="color-input__alpha" max="100" min="0" type="range" value="0">
        </div>
        <div class="color-input__selected-color" title="change format">
          <div class="color-input__crossed-out"></div>
        </div>
      </div>
      <div class="color-input__shade">
        <div class="color-input__pointer"></div>
      </div>
    </div>
  </mock:shadow-root>
</tf-color-input>
`;
